NeoML/src/TraditionalML/SMOptimizer.cpp

```diff
@@ -241,202 +104,383 @@ void CKernelCache::lruInsert(CList *l)
+	swap( y[i], y[j] );
+	swap( d[i], d[j] );
 }
 
+
 //---------------------------------------------------------------------------------------------------
 
-CSMOptimizer::CSMOptimizer(const CSvmKernel& kernel, const IProblem& _data, double _errorWeight, double tolerance, int cacheSize) :
+CSMOptimizer::CSMOptimizer(const CSvmKernel& kernel, const IProblem& _data,
+		int _maxIter, double _errorWeight, double _tolerance, bool _shrinking, int cacheSize) :
 	data( &_data ),
+	maxIter( _maxIter ),
 	errorWeight( _errorWeight ),
-	tolerance(tolerance),
+	tolerance( _tolerance ),
+	shrinking( _shrinking ),
 	Q( FINE_DEBUG_NEW CKernelMatrix( _data, kernel, cacheSize ) ),
-	log(0)
+	log( nullptr ),
+	l( data->GetVectorCount() ),
+	y( Q->GetBinaryClasses() ),
+	QD( Q->GetDiagonal() )
 {
+	weightsMultErrorWeight.SetBufferSize( l );
+	for( int i = 0; i < l; ++i ) {
+		weightsMultErrorWeight.Add( data->GetVectorWeight( i ) * errorWeight );
+	}
+	C = weightsMultErrorWeight.GetPtr();
 }
 
 CSMOptimizer::~CSMOptimizer() 
 { 
 	delete Q;
 }
 
-void CSMOptimizer::Optimize( CArray<double>& alpha, float& freeTerm )
+void CSMOptimizer::Optimize( CArray<double>& _alpha, float& freeTerm )
 {
-	CArray<double> gradient;
-	gradient.Add(-1., data->GetVectorCount() ); // the target function gradient
-	alpha.Empty();
-	alpha.Add( 0., data->GetVectorCount() ); // the support vectors coefficients
+	gradient.Empty();
+	gradient.Add( -1., l ); // the target function gradient
+	g = gradient.GetPtr();
+	gradient0.Empty();
+	gradient0.Add( 0., l ); // gradient, if we treat free variables as 0
+	g0 = gradient0.GetPtr();
+	alphaStatusArray.Empty();
+	alphaStatusArray.Add( AS_LowerBound, l );
+	alphaStatus = alphaStatusArray.GetPtr();
+
+	if( shrinking ) {
+		// shrinking does some permutations in coefficients so use internal alpha array
+		alphaArray.Empty();
+		alphaArray.Add( 0., l ); // the support vectors coefficients
+		alpha = alphaArray.GetPtr();
+		activeSetArray.SetSize( l );
+		activeSet = activeSetArray.GetPtr();
+		for( int i = 0; i < l; ++i ) {
+			activeSet[i] = i;
+		}
+		unshrink = false;
+	} else {
+		_alpha.Empty();
+		_alpha.Add( 0., l );
+		alpha = _alpha.GetPtr();
+	}
+	activeSize = l;
 
-	int maxIter = max( 10000000, data->GetVectorCount() > INT_MAX / 100 ? INT_MAX : 100 * data->GetVectorCount() );
 	int t;
-	for(t = 0; t < maxIter; t++) {
-		// log progress
-		if(log != 0 && t % 1000 == 0) {
-			*log << ".";
+	int counter = min( l, 1000 ) + 1;
+	for( t = 0; t < maxIter; ++t ) {
+		if( --counter == 0 ) {
+			counter = min( l, 1000 );
+			if( shrinking ) {
+				shrink();
+			}
+			// log progress
+			if( log != nullptr ) {
+				*log << ".";
+			}
 		}
-		// Find a pair of coefficients that violate Kuhn - Tucker conditions most of all
+
 		int i, j; 
-		double Gmax, Gmin;
-		findMaximalViolatingPair( alpha, gradient, i, Gmax, j, Gmin);
-		if(Gmax - Gmin < tolerance)	{
-			break;
+		if( selectWorkingSet( i, j ) ) {
+			reconstructGradient();
+			activeSize = l;
```

Comment:
*Discussable*
`activeSize = l;` always goes immediately after `reconstructGradient();`. Does gradient reconstruction always reset the active set? And if so, it's better to move `activeSize = l;` from here (and from one more call) to `reconstructGradient()`